### ISSUE TITLE
PL310 cache opeartions can't take effect on non-secure cacheline if it's specific to the PA

### ISSUE THREAD

**Issue by tiger20081015** [OP-TEE/optee_os#4783]
Hi, we recently met an issue in our platform(CA9 with PL310 enabled): 
We have a hardware crypto IP in secureos. When the IP DMA's output buffer is a dynamic shared memory allocated from REE with non-secure attribution,  the cpu always gets the output buffer's old data even if we have already done the cache invalidate operation of the ouput buffer's physical address before.  If PL310 is  disabled or the memory attribution is changed from non-secure to secure, the issue disappears. 

We found the explnation in the PL310's spec(DDI0246H_l2c310_r3p3_trm.pdf) as below:
![image](https://user-images.githubusercontent.com/35290009/128447317-c25dd6f5-dc37-4592-abaf-daea8f64d00e.png)

From the spec, it seems in secureos the PL310 cache maintanance can't take effect to non-secure memory if it's specfic to PA. In secureos there are several types of non-secure memory that are impacted by this feature , like MEM_AREA_RAM_NSEC, MEM_AREA_SHM_VASPACE.  Do you have any mind to solve this issue? The set/way solution is likely to have performance impact and PL310 seems has no configuration to change its behavior.  Thanks!

**Comment by tiger20081015** [comment]
hi, @etienne-lms , I noticed you submitted the PL310 related patches. Could you please give some advice about solving the problem? Thanks in advance!

**Comment by Emantor** [comment]
The only way I see to get around this problem is to use a secure world buffer for DMA operations and copying the data after the crypto IP is finished. This way you don't DMA into a buffer which you can't operate with cache maintenance ops on. Obviously this will have a performance impact.

**Comment by tiger20081015** [comment]
@Emantor, thanks for your reply. For the issue I met, as you metioned, using a secure buffer to copy data is donable. But since non-secure types of memory like MEM_AREA_RAM_NSEC, MEM_AREA_SHM_VASPACE all have the same problem, it seems quite difficult to find all the places using these memory and apply the copy buffer solution?

**Comment by etienne-lms** [comment]
Hi @tiger20081015,
I think @Emantor suggested a reliable scheme: in secure world, the operation using DMAs would performed using a temporary secure buffer and once DMA is done, secure world copies (after cache sync) the output secure data to the target output shared memory. This should be needed only for the very specific operations (in OP-TEE) where you use DMA to fill non-secure memory, it would not need changes at all places where OP-TEE uses non-secure memory.

**Comment by tiger20081015** [comment]
@etienne-lms I see, thanks for your advice. BTW, I searched the latest optee code and found below 2 places which may have risk when PL310 is enabled. Could you please check if they should be fixed?

1: core/pta/tests/invoke.c
```
333 static TEE_Result test_dump_sdp(uint32_t type, TEE_Param p[TEE_NUM_PARAMS])
334 {
...
353 >-------if (!core_vbuf_is(CORE_MEM_SDP_MEM, src, sz) ||
354 >-------    !core_vbuf_is(CORE_MEM_NON_SEC, dst, sz)) {
355 >------->-------DMSG("bad memref secure attribute");
356 >------->-------return TEE_ERROR_BAD_PARAMETERS;                                                                                                                                                        
357 >-------}
...
367 >-------if (cache_operation(TEE_CACHEFLUSH, dst, sz) != TEE_SUCCESS)  //The dst buffer is non-seccure memory
368 >------->-------return TEE_ERROR_GENERIC;
369 
370 >-------memcpy(dst, src, sz);
371 
372 >-------if (**cache_operation(TEE_CACHEFLUSH, dst, sz)** != TEE_SUCCESS)
373 >------->-------return TEE_ERROR_GENERIC;
...
}
```
2: core/arch/arm/tee/svc_cache.c
 12 TEE_Result syscall_cache_operation(void *va, size_t len, unsigned long op)
 13 {
  ...
 23 >-------/*
 24 >------- * TAs are allowed to operate cache maintenance on TA memref parameters
 25 >------- * only, not on the TA private memory.
 26 >------- */
 27 >-------if (vm_buf_intersects_um_private(&utc->uctx, va, len))
 28 >------->-------return TEE_ERROR_ACCESS_DENIED;
 ...
 37 >-------return cache_operation(op, va, len); //Is it possible that va points to non-secure memory?
 38 }

**Comment by tiger20081015** [comment]
@etienne-lms , Could you please check if the code pasted in above reply be fixed or not?
